codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -358,18 +357,23 @@ public ChannelFuture writeRstStream(ChannelHandlerContext ctx, int streamId, lon
      */
     @Override
     public ChannelFuture writeGoAway(ChannelHandlerContext ctx, int lastStreamId, long errorCode, ByteBuf debugData,
-            ChannelPromise promise) {
-        Http2Connection connection = connection();
-        if (connection.isGoAway()) {
+                                     ChannelPromise promise) {
+        try {
+            if (connection().goAwaySent() && connection().remote().lastKnownStream() < lastStreamId) {
+                throw connectionError(PROTOCOL_ERROR, "Last stream identifier must not increase between " +
+                                                      "sending multiple GOAWAY frames (was '%d', is '%d').",
+                                                      connection().remote().lastKnownStream(),
+                                                      lastStreamId);
+            }
+            connection().goAwaySent(lastStreamId);
```

Comment:
> we really just have to make sure the write occurs before we notify listeners.
+1

> If that's the case, couldn't we just move the goAwaySent to right after the call to write (probably even before we flush)?

The issue is what `goAwaySent(..)` does under the hood.  It is notifying listeners and closing the streams.  I think in the case of `GOAWAY` I think we also need to be able to re-set the `lastStreamId` if the write fails.